equational_theories/EquationsCommand.lean

```diff
@@ -53,7 +73,20 @@ elab mods:declModifiers tk:"equation " i:num " := " tsyn:term : command => do
       | some idx => `(FreeMagma.Leaf $(quote idx.val))
       | none => pure s
   let mut tl : Term := ⟨tl⟩
-  elabCommand (← `(command| abbrev%$tk $lawIdent : Law.MagmaLaw Nat := $tl))
+  let freeMagmaSize := Syntax.mkNumLit (toString is.size)
+  -- define law over `Fin n`
+  elabCommand (← `(command| abbrev%$tk $finLawIdent : Law.MagmaLaw (Fin $freeMagmaSize) := $tl))
+  -- compatibility between the `finLaw` and the original equation
+  let modelsIffLemma : Ident := mkIdent (.mkSimple s!"models_iff_{is.size}")
+  elabCommand (← `(command| abbrev%$tk $finThmName : ∀ (G : Type _) [$inst : Magma G], G ⊧ $finLawIdent ↔ $eqIdent G := $modelsIffLemma $finLawIdent))
+  -- define the actual law over `Nat`
+  elabCommand (← `(command| abbrev%$tk $lawIdent : Law.NatMagmaLaw := $tl))
+  -- compatibility between the law and the original equation
+  elabCommand (← `(command| abbrev%$tk $thmName : ∀ (G : Type _) [$inst : Magma G], G ⊧ $lawIdent ↔ $eqIdent G :=
+                    fun G _ ↦ Iff.trans (Law.satisfies_fin_satisfies_nat G $finLawIdent).symm ($finThmName G)))
```

Comment:
It might be better to do this in a more low-level way, rather than re-elaborating the law repeatedly. A hybrid approach would be to elaborate the law once and use `Expr.toSyntax` to repeat it.

Reply:
Thanks for the suggestion. The law happens to be elaborated slightly differently in the two situations -- in one case, the leaves are coerced to `Fin n`, while in the second case, they are left as natural numbers, so storing the rest of elaboration as an expression may not directly work. (It's possible that I may be misunderstanding the comment.)